### PR TITLE
audit: Make `--display-failures-only` the default for Casks

### LINKED ISSUE #15074: Remove `brew audit --display-failures-only`
### Verification

- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### Provide a detailed description of the proposed feature

`brew audit --display-failures-only` should be `odeprecated`, made the default behaviour and omitting it replaced with `--strict` (or, if appropriate, `--git`, `--online`, `--new-cask`, etc.)

### What is the motivation for the feature?

It's confusing to have a behaviour like this that applies only for casks when there's an equivalent already there for formulae. Similarly, the default output without `--display-failures-only` is very hard to follow as it's hard to find the actual error over all the warnings.

### How will the feature be relevant to at least 90% of Homebrew users?

Improved contributor and maintainer productivity.

### What alternatives to the feature have been considered?

Doing nothing.

### PR DESCRIPTION
**Pull request by issyl0**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

- Fixes #15074.
- Deprecate `brew audit --display-failures-only` for Casks. The default behaviour for formulae is to only display failures, so let's make Casks behave the same way since all the warnings are noisy.
- Only show warnings (less important errors, I guess, now) for Casks if `--strict` (or `--new-cask` which implies `--strict`) is used in `brew audit`.

### CODE REVIEW

File: Library/Homebrew/cask/audit.rb [hunk 1]
```diff
@@ -76,66 +76,37 @@ def errors
...
-      !(errors? || warnings?)
+      !errors?
     end
 
-    sig { params(message: T.nilable(String), location: T.nilable(String)).void }
-    def add_error(message, location: nil)
-      errors << ({ message: message, location: location })
-    end
+    sig { params(message: T.nilable(String), location: T.nilable(String), warning: T::Boolean).void }
+    def add_error(message, location: nil, warning: false)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/audit.rb:90]: IMO
```suggestion
    def add_error(message, location: nil, strict_only: false)
```
or
```suggestion
    def add_error(message, location: nil, stricty: false)
```
would make more sense?

File: Library/Homebrew/cask/auditor.rb [hunk 2]
```diff
@@ -26,7 +26,6 @@ def initialize(
       quarantine: nil,
       any_named_args: nil,
       language: nil,
-      display_passes: nil,
       display_failures_only: nil,
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/auditor.rb:29]:
```suggestion
```

will presumably be able to go away. 

I'd suggest having a `strict` here somewhere, too.

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good so far!

File: Library/Homebrew/cask/auditor.rb (hunk 2 above)
Comment (issyl0) [reply Library/Homebrew/cask/auditor.rb:29]: `@audit_strict` already exists.

File: Library/Homebrew/cask/audit.rb [hunk 3]
```diff
@@ -71,66 +71,38 @@ def errors
...
     def success?
-      !(errors? || warnings?)
+      !errors?
     end
 
-    sig { params(message: T.nilable(String), location: T.nilable(String)).void }
-    def add_error(message, location: nil)
-      errors << ({ message: message, location: location })
-    end
+    sig { params(message: T.nilable(String), location: T.nilable(String), strictish: T::Boolean).void }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/audit.rb:84]: Why `strictish` instead of `strict` or `strict_only`?

File: Library/Homebrew/cask/auditor.rb [hunk 4]
```diff
@@ -63,23 +58,19 @@ def audit
...
         audit = audit_cask_instance(cask)
-        summary = audit.summary(include_passed: output_passed?, include_warnings: output_warnings?)
-        puts summary if summary.present? && output_summary?(audit)
-        warnings += audit.warnings
+        puts audit.summary if audit.summary.present? && output_summary?(audit)
         errors += audit.errors
       end
 
-      { warnings: warnings, errors: errors }
+      { errors: errors }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/auditor.rb:73]:
```suggestion
      errors
```
may be a better API? Up to you.

File: Library/Homebrew/dev-cmd/audit.rb [hunk 5]
```diff
@@ -65,7 +65,7 @@ def self.audit_args
              description: "Prefix every line of output with the file or formula name being audited, to " \
                           "make output easy to grep."
       switch "--display-failures-only",
-             description: "Only display casks that fail the audit. This is the default for formulae."
+             description: "Only display casks that fail the audit. This is the default for formulae and casks."
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb:68]:
```suggestion
             description: "Only display casks that fail the audit. This is the default for formulae and casks.",
             hidden: true
```

IMO if we're deprecating it

**Review by MikeMcQuaid — COMMENTED** [review]
Great work! Almost there, a few more thoughts.

**Review by MikeMcQuaid — APPROVED** [review]
Thanks @issyl0, great work!

**Comment by MikeMcQuaid** [comment]
Looks like `cask audit` is failing but everything else looks 🎉 https://github.com/Homebrew/brew/actions/runs/4610240508/jobs/8150188692?pr=15105#step:7:18

**Review by MikeMcQuaid — APPROVED** [review]

File: Library/Homebrew/cask/audit.rb [hunk 6]
```diff
@@ -663,16 +641,9 @@ def audit_github_repository_archived
...
       return unless metadata["archived"]
 
-      message = "GitHub repo is archived"
-
-      if cask.discontinued?
-        add_warning message
-      else
-        add_error message
-      end
+      add_error("GitHub repo is archived", strict_only: cask.discontinued?)
```
Comment (reitermarkus) [inline Library/Homebrew/cask/audit.rb:646]: Shouldn't this be `!cask.discontinued?`? Naming this `strict_only` somehow made it more confusing.

If the cask is discontinued, I don't think this should ever be an error.

**Review by reitermarkus — CHANGES_REQUESTED** [review]

File: Library/Homebrew/cask/audit.rb (hunk 6 above)
Comment (issyl0) [reply Library/Homebrew/cask/audit.rb:646]: I preferred the old name too. 😬 

The `strict_only` naming is what used to be a warning. So if the cask is discontinued then there'll be a message show up if `brew audit --strict` is run.

I think this is the same behaviour as before (if you look at the lines before in the diff and translate the logic to the new names).

Comment (reitermarkus) [reply Library/Homebrew/cask/audit.rb:646]: But now we only have errors, no warnings, right? So I don't think this should be an error, even with `--strict`.

Comment (issyl0) [reply Library/Homebrew/cask/audit.rb:646]: Ohhhh. Right.

Comment (issyl0) [inline Library/Homebrew/cask/audit.rb:646]:
```suggestion
      # Discontinued casks shouldn't show up as errors.
      add_error("GitHub repo is archived", strict_only: !cask.discontinued?)
```

File: Library/Homebrew/cask/audit.rb [hunk 7]
```diff
@@ -663,16 +641,10 @@ def audit_github_repository_archived
...
 
-      message = "GitHub repo is archived"
-
-      if cask.discontinued?
-        add_warning message
-      else
-        add_error message
-      end
+      # Discontinued casks shouldn't show up as errors.
+      add_error("GitHub repo is archived", strict_only: !cask.discontinued?)
```
Comment (reitermarkus) [inline Library/Homebrew/cask/audit.rb:647]:
```suggestion
      # Discontinued casks shouldn't show up as errors.
      add_error("GitHub repo is archived") unless cask.discontinued?
```

File: Library/Homebrew/cask/audit.rb [hunk 8]
```diff
@@ -684,16 +656,9 @@ def audit_gitlab_repository_archived
...
       return unless metadata["archived"]
 
-      message = "GitLab repo is archived"
-
-      if cask.discontinued?
-        add_warning message
-      else
-        add_error message
-      end
+      add_error("GitLab repo is archived", strict_only: cask.discontinued?)
```
Comment (reitermarkus) [inline Library/Homebrew/cask/audit.rb:661]:
```suggestion
      # Discontinued casks shouldn't show up as errors.
      add_error("GitLab repo is archived") unless cask.discontinued?
```

**Review by reitermarkus — CHANGES_REQUESTED** [review]

**Comment by reitermarkus** [comment]
Effectively, an archived repo implies that the cask is discontinued. So adding a `discontinued` should fix the audit.

**Comment by issyl0** [comment]
Thanks Markus. Sorry, on my phone, didn't actually test that it worked. 😅

**Review by reitermarkus — APPROVED** [review]

**Comment by issyl0** [comment]
Closing and re-opening to re-trigger the "expected" CI checks.

**Review by MikeMcQuaid — APPROVED** [review]
